src/messages_manager.ts

```diff
@@ -8,78 +10,190 @@ import { disposeAll } from "./utils";
+
+            default:
+                return;
         }
+
+        switch (name) {
+            case "msg_clear":
+            case "msg_history_clear":
+                // These clear messages are often followed by a flush whenever neovim
+                // thinks it's "done" showing those messages, resulting in an empty output
+                // panel instead of the desired display. To avoid flushing the now-empty
+                // buffer, we skip setting didChange so the flush becomes a no-op.
+                // Seems likely caused by/related to https://github.com/neovim/neovim/issues/20416
+                break;
+
+            default:
+                this.didChange = true;
+        }
+
+        logger.trace(name, inspect(args, { depth: 5, compact: 3 }));
     }
 
-    private writeMessage(msg: string): void {
-        if (msg.length === 0) {
-            return;
+    private async handleFlush(): Promise<void> {
+        if (!this.didChange) return;
+
+        const messages = this.displayHistory ? this.historyBuffer : this.messageBuffer;
+        logger.trace(`Flushing ${this.displayHistory ? "history" : "message"} buffer: ${inspect(messages)}`);
+
+        const msg = messages.join("\n");
+
+        const lineCount = msg.split("\n").length;
+        // Before Nvim 0.10, cmdheight is unchangeable, and it's always 0.
+        const cmdheight = Math.max(1, (await this.main.client.getOption("cmdheight")) as number);
+
+        const revealForCmdheight = lineCount > cmdheight;
+        if (revealForCmdheight && cmdheight < 2) {
+            this.warnCmdheightRevealBehavior(cmdheight);
         }
 
-        logger.info(msg);
-        const outputMsg = this.ensureEOL(msg);
-        this.channel.replace(outputMsg);
+        const shouldRevealOutput = this.revealOutput || revealForCmdheight;
+        const { didChange, displayHistory } = this;
+        logger.trace(inspect({ didChange, shouldRevealOutput, displayHistory, lineCount }));
+
+        this.writeMessage(this.ensureEOL(msg));
+        if (shouldRevealOutput) {
+            this.channel.show(true);
+        }
+
+        // Reset all the state for the next batch of redraw messages
+        this.didChange = false;
+        this.displayHistory = false;
+        this.revealOutput = false;
+    }
+
+    private async warnCmdheightRevealBehavior(cmdheight: number): Promise<void> {
+        if (this.context.globalState.get(SUPPRESS_CMDHEIGHT_MSG)) return;
+
+        const suppress = "Don't show again";
+        const picked = await window.showInformationMessage(
+            `The vscode-neovim output panel will be revealed for output longer than
+            ${cmdheight} line(s). To suppress output for messages from searching with /,
+            use neovim version 0.10+ and configure cmdheight=2 or greater.`,
+            "Ok",
+            suppress,
+        );
+
+        if (picked === suppress) {
+            this.context.globalState.update(SUPPRESS_CMDHEIGHT_MSG, true);
```

Comment:
can we just bump our required Nvim version? that eliminates this code and avoids noise in globalState.